### PR TITLE
[TE] Support mixing normal and cross-thread reduction

### PR DESCRIPTION
**Pull request by roastduck**
Currently TVM only supports pure normal (i.e. sequential) reduction or pure cross-thread reduction. Since TVM has not supported nested reduction yet, one is even unable to schedule a mixed reduction manually. I modified the function that lowers cross-thread reduction, to support mixed reduction as well.

The approach is straight forward: First perform normal reduction into local variables in each threads first, and then invoke the original cross-thread reduction intrinsic. It works like this (pseudo-code):

```c++
// Divide the loop nest into two parts
normal_red = sequantial loops nest
common = other loops nest
    
// If normal_red is empty, fallback to original code

normal_init = generate init for the temp var
normal_update = generate sequential reduction on the temp var
body genereate cross-thread reduction // original code

// Merge loop nests and add some checks
body = SeqStmt(normal_init, MergeNest(normal_red, normal_update), body)
body = MergeNest(common, body)
return body

```

A test case is added as a Python unit test.

This is my first PR to TVM, and I am not sure whom to invite as a reviewer. Since this is compiling pass related, @tqchen can you review my code?

### CODE REVIEW

**Comment by tqchen** [comment]
Thanks @roastduck  for contributing.  cc @vinx13 @merrymercy @Hzfengsy @wpan11nv  please also help to take a look:)

File: src/te/operation/cross_thread_reduction.cc [hunk 1]
```diff
@@ -57,10 +57,63 @@ Stmt MakeCrossThreadReduction(
   for (PrimExpr v : conds) {
     cond = cond && v;
   }
+
+  std::vector<std::vector<Stmt> > common, normal_red;
```
Comment (wpan11nv) [inline src/te/operation/cross_thread_reduction.cc:61]: no space before > now (> C++11).

**Comment by roastduck** [comment]
All resolved. @wpan11nv

**Comment by wpan11nv** [comment]
LGTM. Thanks.

**Comment by tqchen** [comment]
Thanks @roastduck @wpan11nv !

**Pull request merged**